[Feature #1] 프로젝트 기본 세팅 & 로그인(간략화 버전) API 구현

isExample opened the pull request:
## ✏️ 작업 요약
Spring Boot 프로젝트 개발에 필요한 기본적인 환경을 구성하고 간단한 형태로 로그인을 구현했습니다.

## ✨ 작업 내용 및 리뷰 포인트
- MySQL Datetime vs Timestamp
Datetime에 비해 더 적은 저장 공간이 필요하고 여러 시간대에서 애플리케이션을 이용하기에 적합합니다.
또한, 데이터 저장 시 바이너리 형태로 저장이 되기 때문에 인덱싱 측면에서도 더 높은 성능을 기대할 수 있습니다.
Unix epoch와 관련된 최대 시간 범위 문제는 MySQL 8.0.28 버전 이후부터 최대 시간 범위가 늘어남에 따라 해결되어 Timestamp로 결정했습니다.

- Mybatis Mapper
Mybatis의 활용도를 높이기 위해 xml 형태의 Mapper 파일을 통해 쿼리문을 관리하여 java 소스 코드와 쿼리문을 분리하고자 했습니다.

## 🔖 이슈 번호
- [x] #1

isExample committed:
[Feature] API 공통 응답 로직 구현

기존 Spring의 ResponseEntity보다 간단하게 응답을 생성하고 원하는 값들을 포함시키기 위해 커스텀했습니다.

isExample committed:
[Feature] 예외 처리를 위한 ExceptionAdvice 및 GeneralHandler 구현

실패 응답 구조에 맞춰 RuntimeException 및 그 하위 예외들을 처리합니다.

isExample committed:
[Refactor] IntelliJ의 기본 코드 스타일에 맞게 수정

IntelliJ 기본 코드 스타일에 맞게 수정하고 자동 EOF를 설정했습니다.

isExample committed:
[Feature] Logback 로깅 환경 구성

Logback 환경 및 profile을 구성하여 운영 환경에 따른 로깅을 설정했습니다.

isExample committed:
[Refactor] e.printStackTrace()를 e.getMessage()로 대체

e.printStackTrace()가 보안적으로 취약하고 로깅 파일에 기록되지 않아 e.getMessage()로 수정했습니다.

isExample committed:
[Feature] 테이블 초기화 및 더미 데이터 스크립트 작성

테이블들을 초기화하고 테스트용 더미 데이터를 하나씩 추가하는 스크립트를 작성했습니다.

isExample committed:
[Feature] Schema ID 체계 및 금액 타입 수정

ID 체계를 Auto-increment로 설정하고(추후 변경 예정) 금액 타입을 DECIMAL -> INT로 수정했습니다.

isExample committed:
[Feature] 초기 데이터 id 값 미지정

Auto-increment로 수정함에 따라 초기 데이터 insert문에서 id를 지정하지 않도록 합니다.

isExample committed:
[Refactor] 시간 필드 Datetime -> Timestamp

여러 시간대에서 액세스하기 용이하고 저장 효율성이 좋은 Timestamp로 타입을 변경했습니다.

isExample committed:
[Feature] Database Config 및 의존성 추가

Mybatis 매핑을 위해 Config 클래스를 구현하고 의존성을 추가했습니다.

isExample committed:
[Feature] Member 테이블 관련 객체와 매퍼 생성

Member 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Product 테이블 관련 객체와 매퍼 생성

Product 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Orders 테이블 관련 객체와 매퍼 생성

Orders 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Orders 테이블의 status 기본값 설정

status의 기본값을 PENDING(대기)로 설정했습니다.

isExample committed:
[Feature] OrderDetail 테이블 관련 객체와 매퍼 생성

Order_Detail 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Cart 테이블 관련 객체와 매퍼 생성

Cart 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Payment 테이블 관련 객체와 매퍼 생성

Payment 테이블의 데이터에 대응되는 객체를 추가하고 매핑했습니다.

isExample committed:
[Feature] Swagger 의존성과 Config 추가

Swagger UI를 생성하기 위해 의존성과 기본 Swagger Config를 추가했습니다.

isExample committed:
[Refactor] 단일 행 조회시 Optional 활용

select 쿼리로 불러온 데이터가 null인 경우를 대비해서 Optional을 사용합니다.

isExample committed:
[Feature] 로그인(간략화) 기능을 위한 DTO와 controller 구현

로그인 기능을 위한 DTO와 controller를 구현했습니다.
로그인 정보 유효성 검사를 위해 validation 의존성과 코드를 추가했습니다.

isExample committed:
[Feature] 로그인(간략화) 기능을 위한 mapper와 service 구현

로그인 기능을 위한 mapper와 service를 구현했습니다.
HttpSession의 의존성을 service에 두지 않기 위해 controller단에서 세션 등록을 진행합니다.

flab-true commented on build.gradle:
```diff
@@ -24,6 +24,9 @@ repositories {
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:3.0.3'
```
io.spring.dependency-management 플러그을 사용하여 Spring Boot의 버전에서 spring-boot-dependencies bom 을 거져 오도록 작성할 수 있습니다. 개선을 검토해 주시면 좋겠습니다. 
-  https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/htmlsingle/
- https://hsik0225.github.io/spring/2021/04/17/Spring-Dependency-Management/

flab-true commented on src/main/java/com/flab/order/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+            session.setAttribute("memberEmail", request.getEmail());
+            return ApiResponse.onSuccess("로그인에 성공하였습니다.");
+        }
+        throw new GeneralHandler(ErrorStatus.INVALID_PASSWORD);
```
정상적인 케이스에 throw new를 하는것 보다는 아래와 같이 코드를 개선하는 방향은 어떠세요?
```java 
        if (!memberService.authenticate(request.getEmail(), request.getPassword());) {
                throw new GeneralHandler(ErrorStatus.INVALID_PASSWORD);
        }

        session.setAttribute("memberEmail", request.getEmail());
        return ApiResponse.onSuccess("로그인에 성공하였습니다.");
```

flab-true commented on src/main/java/com/flab/order/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+
+    @Operation(summary = "로그인 API")
+    @PostMapping("/login")
+    public ApiResponse<String> login(@Valid @RequestBody MemberRequest.Login request, HttpSession session) {
```
`@validated`와  `@valid` 차이는 무엇일까요? 
주석으로 달아놓으셔도 좋은 방법이 될 것 같아요.

flab-true replied:
각각 `annotation`에서 발생하는 `Exception`을 포함해서 정리해 주시면 좋겠습니다.

isExample replied:
@Valid는 `자바 표준 스펙`으로, `Bean Validator`를 통해 유효성을 검증하는 어노테이션입니다.
Bean Validator의 특징에 따라 객체 필드에 달린 어노테이션을 통해 간편하게 검증할 수 있습니다.

@Valid는 Argument Resolver에 의해 처리됩니다. 제가 @Valid를 사용한 @RequestBody의 경우, 그 구현체인 RequestResponseBodyMethodProcessor를 통해 유효성 검증이 진행됩니다. 검증에 오류가 발생하면 MethodArgumentNotValidException를 던지고 디스패처 서블릿에 기본으로 등록된 DefaultHandlerExceptionResolver에 의해 400 BadRequest 에러를 발생시킵니다. 이러한 이유로 @Valid는 기본적으로 컨트롤러에서만 동작합니다.

@Validated는 자바 표준 스펙이 아닌 `스프링에서 제공하는 어노테이션`입니다.
`AOP 기반`으로 메소드의 요청을 가로채어 유효성을 검증합니다. Argument Resolver를 통한 검증이 아니기 때문에 계층과 무관하게 사용할 수 있어, controller 계층이 아닌 곳에서 검증이 필요할 때 사용할 수 있습니다. 검증에 오류가 발생하면 ConstraintViolationException을 던집니다.

isExample replied:
현재는 @Valid만 이용하고 있어서 이후에 @Validated를 이용하게 된다면 주석을 추가해보도록 하겠습니다!

flab-true commented on src/main/java/com/flab/order/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+    @Operation(summary = "로그인 API")
+    @PostMapping("/login")
+    public ApiResponse<String> login(@Valid @RequestBody MemberRequest.Login request, HttpSession session) {
+        boolean isAuthenticated = memberService.authenticate(request.getEmail(), request.getPassword());
```
정상적인 케이스에 throw new를 하는것 보다는 아래와 같이 코드를 개선하는 방향은 어떠세요?
```java 
        if (!memberService.authenticate(request.getEmail(), request.getPassword());) {
                throw new GeneralHandler(ErrorStatus.INVALID_PASSWORD);
        }

        session.setAttribute("memberEmail", request.getEmail());
        return ApiResponse.onSuccess("로그인에 성공하였습니다.");
```

isExample replied:
이는 “Guard Clause”라고도 불리는 클린 코드의 한 형태더군요.
크게 인식하지 못했던 부분인데 참고하여 말씀해 주신 방향으로 개선했습니다!

flab-true commented on build.gradle:
```diff
@@ -24,6 +24,9 @@ repositories {
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:3.0.3'
```
`io.spring.dependency-management` 플러그을 사용하여 Spring Boot의 버전에서 `spring-boot-dependencies bom` 을 가져 오도록 작성할 수 있습니다. 개선을 검토해 주시면 좋겠습니다. 
-  https://docs.spring.io/spring-boot/docs/current/gradle-plugin/reference/htmlsingle/
- https://hsik0225.github.io/spring/2021/04/17/Spring-Dependency-Management/

isExample replied:
`io.spring.dependency-management` 플러그인을 사용하면 Spring Boot로 애플리케이션을 구축할 때 의존성 및 관련 버전 관리를 단순화할 수 있습니다. 각 의존성에 대한 버전을 bom에 명시된 일관된 버전으로 반영하여 수동으로 지정할 필요가 없다는 장점이 있습니다.

하지만 이에 포함되지 않는 의존성들은 직접 버전을 관리해야 합니다. 제 프로젝트에서는 mybatis와 springdoc이 이에 해당되어 해당 의존성은 버전을 명시해 두었습니다. (`gradle dependencyManagement` 명령어를 통해 지원하는 의존성들 목록을 검토해 봤는데 해당되지 않았습니다.)

추가로, 제가 의도적으로 해당 플러그인을 추가한 적은 없는데 설정이 되어 있더군요. Spring Initializr로 프로젝트를 만들게 되면 기본적으로 추가되는 플러그인 듯하여 어떠한 의존성도 추가하지 않은 프로젝트를 테스트겸 생성해보니 실제로 기본적으로 반영되는 플러그인이라는 것을 알 수 있었습니다. 아래는 어떠한 의존성도 추가하지 않은 프로젝트의 build.gradle입니다.

<img width="547" alt="기본 프로젝트 build.gradle" src="https://github.com/f-lab-edu/order-service/assets/93423346/8a07915a-390b-4e9a-95f2-593b436ccf01">

flab-true commented on build.gradle:
```diff
@@ -24,6 +24,9 @@ repositories {
 dependencies {
```
`Gradle warpper`를 구성하지 않았는데요. 알아두시면 좋을 것 같습니다. 
왜 사용해야 할까요?

isExample replied:
Gradle 프로젝트를 빌드하고 실행하기 위해서는 로컬 컴퓨터에 해당 프로젝트와 버전이 동일해야 합니다. 즉, 사용하려는 `프로그램의 Gradle 버전과 동일하게 Gradle을 설치`해주어야 하는데 서로 다른 프로젝트에 대해 매번 이 작업을 수행하기에는 비효율적입니다. 이러한 문제는 Gradle Wrapper를 사용하여 해결할 수 있습니다.

Gradle Wrapper는 Gradle을 미리 설치하지 않고도 누구나 Gradle 프로젝트로 작업할 수 있도록 해주는 스크립트입니다. `지정된 버전의 Gradle을 자동으로 다운로드하고 해당 버전을 사용하여 빌드를 실행`하여 모든 환경과 프로젝트의 모든 개발자 간에 일관성을 보장합니다.

기존에 gradle/wrapper 디렉토리를 비롯해서 관련 스크립트가 모두 존재했지만 wrapper 버전이 누락되어 구성되지 않았던 것 같습니다. gradle 버전을 추가하여 다시 구성했습니다.

flab-true replied:
Docker Image를 만들때 어떻게 작업을 하셨을까요? 빌드 도구에는 build task가 있습니다. 무엇을 하는 용도일까요?

isExample replied:
Docker Image를 만들 때 OpenJDK 이미지를 지정한 후, Gradle로 빌드된 JAR 파일(실행 가능한 JAR)을 Docker 이미지 내부의 지정된 위치로 복사했습니다. 그리고 Docker 컨테이너가 시작될 때 실행되어야 하는 명령어들을 설정해서 작업했습니다.

build task는 실행 파일, 라이브러리, 웹 애플리케이션의 WAR 파일 등을 실행 가능한 형태로 변환하는 과정을 자동화합니다.
일반적으로 build script에 담기며 빌드 과정이 구체적으로 정의됩니다.

flab-true replied:
넵 답변 감사합니다. Dockerfile도 정리해서 올려주시면 좋을 것 같네요.
해당 Image는 Docker Hub 또는 Docker Registry에 업로드하여 실행해보신 경험이 있으신지 궁금합니다.

flab-true commented on src/main/java/com/flab/order/config/SwaggerConfig.java:
```diff
@@ -0,0 +1,24 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class SwaggerConfig {
```
실서버에서 Swagger 문서가 노출될 것 같습니다. 개발에서만 노출되는 형태로 고민해 주세요.

isExample replied:
이 부분도 실서버와 개발 환경 설정을 달리하는 과정에서 시도해보겠습니다!

flab-true commented on src/main/resources/application.properties:
```diff
@@ -1,6 +1,8 @@
 spring.datasource.url=jdbc:mysql://localhost:3306/order
 spring.datasource.username=${DATABASE_USERNAME}
 spring.datasource.password=${DATABASE_PASSWORD}
+spring.sql.init.mode=always
```
실서버에서도 sql이 실행될 것 같습니다. 개발 환경(Eg) dev, qa, stage, prod) 따라 설정 값을 달리 로딩하도록 고려해 주세요.
- application.properties 
- build.gradle

isExample replied:
실서버와 개발 환경 설정 값을 달리 하는 부분은 조금 더 학습해 본 후, 별도의 PR에서 다뤄보도록 하겠습니다. 피드백 감사합니다!

flab-true commented on build.gradle:
```diff
@@ -24,6 +24,9 @@ repositories {
 dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-data-jdbc'
 	implementation 'org.springframework.boot:spring-boot-starter-web'
+	implementation 'org.mybatis.spring.boot:mybatis-spring-boot-starter:3.0.3'
```
아래 의존성도 추가해 주세요. 어떤 목적으로 사용될까요? 주석에 포함해도 좋을 것 같습니다. 
- `org.springframework.boot:spring-boot-devtools`

isExample replied:
`org.springframework.boot:spring-boot-devtools` 의존성은 다음의 기능들을 제공합니다.

- Property Defaults
- Automatic Restart
- Live Reload
- Global Settings
- Remote Applications

서버 측에서는 이 DevTools를 사용하여 개발 효율성을 높일 수 있습니다. 현재 제 개발 상황에서는 변경 ~ 확인 과정에서 소요되는 시간을 줄이는 데에 큰 기여를 할 것 같아 의존성을 추가했습니다!

flab-true commented on src/main/java/com/flab/order/domain/dto/MemberRequest.java:
```diff
@@ -0,0 +1,19 @@
+package com.flab.order.domain.dto;
```
package는 `dto`로 구성되어 있습니다. 
DTO는 원칙적으로 불변성을 가져야하지만, 실제 사용시에는 상황에 따라 변경 가능성을 가질 수도 있습니다. 
`DTO`와 `VO`의 차이는 무엇인가요?

isExample replied:
DTO(Data Transfer Object)와 VO(Value Object)의 주요한 차이점은 `사용 목적`에 있습니다. 

DTO는 계층 간의 데이터 전송을 위한 객체로, 로직을 포함하지 않는 순수한 데이터 구조입니다. 
저는 server와 client 간의 `API의 요청과 응답 과정에서 전해지는 데이터들의 컨테이너` 역할을 위해 dto라는 package에 DTO 클래스들을 구현했습니다. 해당 MemberRequest는 회원과 관련된 요청 데이터에 대한 컨테이너입니다.

반면, VO는 비즈니스 로직을 수행하는 데 사용되는 정보를 캡슐화하는 용도로 사용됩니다. 의미 있는 비즈니스 개념을 표현하면서 불변성을 보장합니다. 즉, 원시값을 포장하여 도메인에서 의미있는 명확한 클래스로 표현할 수 있습니다.

flab-true commented on src/main/java/com/flab/order/domain/dto/MemberRequest.java:
```diff
@@ -0,0 +1,19 @@
+public class MemberRequest {
+
+    @NoArgsConstructor
+    @Getter
```
@Setter를 안하신 의도가 있다고 생각합니다. 이유가 무엇인가요?  `java immutable`에 대해 알아보시면 좋을 것 같습니다

isExample replied:
Java Immutable은 `생성 후 그 상태를 바꿀 수 없는 객체`를 의미합니다. 불변 객체는 다음과 같은 특징들을 가집니다.

- Thread-safe하여 동기화를 고려하지 않아도 된다.
- Side Effect를 피해 오류가능성을 최소화할 수 있다.
- 가비지 컬렉션의 성능을 높일 수 있다.

해당 DTO 클래스에서 @Setter를 작성하지 않은 이유도 불변함과 관련이 있습니다. 

Request DTO는 주로 HTTP 요청에서 수신된 데이터를 처리하기 위해 컨트롤러 메서드의 매개 변수로 사용됩니다. @Setter 메소드를 생략함으로써 이 클래스의 각 인스턴스에 포함된 정보가 `초기화된 후에도 변경되지 않은 상태로 유지되도록 보장`하고자 @Setter는 작성하지 않았습니다.

zion3mx replied:
네 답변 감사합니다. lombok에서 `immutable`한 `Class`는 어떻게 구현하는게 좋을까요?
@Value를 공부해 주시고, 어떤 목적으로 사용되는지 알려주시면 좋겠습니다.

flab-true replied:
네 답변 감사합니다. lombok에서 `immutable`한 `Class`는 어떻게 구현하는게 좋을까요?
`@value`를 공부해 주시고, 어떤 목적으로 사용되는지 알려주시면 좋겠습니다.

flab-true replied:
온라인 멘토링에서 하신 설명을 들어보면 `vo` 개념이 더 맞지 않나 생각합니다. 
`@value` annotation을 학습해주시고 개선하는 방향을 검토해 주세요.

isExample replied:
블로그들을 위주로 VO와 DTO의 차이를 공부했을 때, 제가 접한 대부분이 VO와 DTO의 특징을 아래와 같이 정의하고 있었습니다.
<img width="400" alt="DTO와 VO 비교 테이블" src="https://github.com/f-lab-edu/order-service/assets/93423346/7df6306c-3d86-4afe-a684-5a69955b821a">
![Presentation](https://github.com/f-lab-edu/order-service/assets/93423346/e02e8b16-9727-4be9-b510-24e73ac49b2a)
그래서 저는 “**목적**” 부분에 집중하여 client로부터 받아오는 데이터 객체를 DTO로 정의하는게 맞다고 판단했습니다.

피드백에 따라 추가적인 학습을 진행해 보니 “**RequestBody는 DTO일까? 아니면 VO일까?**” 에 대해 정의가 일관되지 않은 정보들을 많이 접하게 되었습니다. 대부분은 “목적”으로 인해 DTO로 소개되는 듯하지만 “가변성”에 초점을 맞추면 VO가 적합한 것도 사실입니다. 

고민을 많이 해봤는데 기준점을 어디에 두느냐에 따라 정답이 없는 문제 같다는 생각이 들었고 DTO라는 명칭이 Request Body를 더 적절히 소개한다고 생각되어 DTO로 결정하게 되었습니다.

제가 잘못 이해한 부분이 있다면 피드백 남겨주시면 감사하겠습니다!

isExample replied:
공식 문서에서는 @Value를 다음과 같이 소개합니다.
> @Value is the immutable variant of @Data

Lombok 라이브러리의 @Value는 클래스를 불변 객체로 만들기 위해 사용되며 다음과 같은 기능들을 제공합니다.

- 해당 클래스의 모든 필드를 자동으로 final과 private으로 설정한다. 
- 모든 인자를 받는 생성자를 만든다. (기본 생성자는 만들어지지 않음)
- 모든 필드에 대해 Getter 메서드를 만든다.
- equals(), hashCode(), toString() 메서드를 만든다.

즉, lombok @Value를 활용해서 immutable object를 만들 수 있습니다.
MemberRequest.Login DTO도 이를 이용해서 명시적으로 Immutable하게 개선했습니다.

flab-true replied:
lombok @value를 활용한 Immutable 개선 작업 감사합니다. 
> 즉, lombok https://github.com/value를 활용해서 immutable object를 만들 수 있습니다.
> MemberRequest.Login DTO도 이를 이용해서 명시적으로 Immutable하게 개선했습니다.

flab-true commented on src/main/java/com/flab/order/global/response/statusEnums/ErrorStatus.java:
```diff
@@ -0,0 +1,43 @@
+    _UNAUTHORIZED(HttpStatus.UNAUTHORIZED, "COMMON401", "인증이 필요합니다."),
+    _INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "COMMON500", "서버 내의 문제가 발생했습니다."),
+
+    MEMBER_NOT_EXIST(HttpStatus.BAD_REQUEST, "MEMBER4000", "존재하지 않는 회원 정보입니다."),
```
회원 정보가 없을 때 400을 사용하셨습니다. 400을 사용하신 이유를 설명해 주시면 좋겠습니다.

isExample replied:
경험적으로 특정 id를 통해 레코드를 조회하지만 데이터가 없는 예외는 대부분 다음의 상황에서 발생했습니다.
- API: /members/1 - id값이 1인 회원을 조회한다.
- 하지만 id값이 1인 회원이 존재하지 않는다.

이러한 경우에 client 측에서 보낸 “/members/1”를 HTTP 상태 코드 400번에 해당되는 “malformed request syntax”로 보고 400으로 선택했었습니다.

하지만, 해당 부분은 API url 과 직접적인 관계가 없고 고민해봤을 때 구문적인 오류가 아니라는 점에서 400보다는 404 Not Found, “The server cannot find the requested resource.”가 더 적합한 것 같아 404로 수정했습니다.

flab-true commented on src/main/java/com/flab/order/interceptor/AuthorizationInterceptor.java:
```diff
@@ -0,0 +1,22 @@
+import org.springframework.web.servlet.HandlerInterceptor;
+
+@Component
+public class MemberAuthorizationInterceptor implements HandlerInterceptor {
```
최초 로그인 요청에서 해당 인터셉터가 동작하여 `GeneralHandler(ErrorStatus._UNAUTHORIZED)` 예외가 발생될 것 같은데요. 로그인 요청에는 인터셉터를 거치지 않도록 설정이 필요한 건 아닌지 궁금합니다.

isExample replied:
네, 말씀하신 대로 인터셉터를 거치지 않도록 경로를 설정해야 합니다.
해당 파일은 다른 구현 파일을 커밋 하는 과정에서 실수로 올라간 것이라, 제거하고 완성된 코드를 올려 두었습니다..!

flab-true commented on src/main/java/com/flab/order/mapper/MemberMapper.java:
```diff
@@ -0,0 +1,12 @@
+
+@Mapper
+public interface MemberMapper {
+    Optional<Member> selectMemberById(Long id);
```
네이밍에 대한 코멘트입니다. 절대적인 기준은 아니며, 개발팀의 선호나 프로젝트의 컨텍스트에 따라 기준은 다르다고 생각합니다.
보통 특정 식별자로 DB의 레코드를 가져오는 경우는 `find`가 적절한 것 같습니다.(JPA 사례) 일반적인 검색에서 `select`를 사용하곤 합니다. 
지금의 경우는 id, email를 지정하여 특정 레코드를 찾기 때문에 `findBy...`가 명확한 네이밍이라고 생각합니다. 개인적으로 `Member` 라벨은 MeberMapper에 종속적이라 제외해도 되지 않나 싶네요.
- `selectMemberById` -> `findById`
- JPA Query Methods 스타일을 공부해보세요. https://docs.spring.io/spring-data/jpa/reference/jpa/query-methods.html

isExample replied:
MyBatis 레퍼런스 자료들에서 대부분 `select~` 의 형태로 사용하여 JPA에 비해 개발자가 직접적으로 SQL 작업을 한다는 점에서 직관적인 `select~` 형태를 사용하는 줄 알았습니다. 피드백에 따라 찾아본 결과, MyBatis는 자체적으로 규정한 명명 규칙은 없더군요. 

말씀해 주신 대로 개인의 선호도나 팀 내 규칙에 따라 결정할 수 있을 것 같습니다. 저 역시도 특정 식별자를 통해 특정 레코드를 조회한다는 점에서 find~가 더 적절하다고 생각됩니다. (JPA가 익숙한지라 더 자연스럽게 느껴지기도 합니다.) 

따라서, `select~`를 `findBy~`로 수정하고 중복되는 객체 키워드(Member 등)를 제거하는 형태로 개선했습니다!

flab-true reviewed (commented):
코멘트 확인 감사합니다. :-)

isExample committed:
[Refactor] Guard Clause를 따라 오류 먼저 처리

코드 블록의 중첩을 최소화하고 성공 흐름을 더 명확하게 표현할 수 있도록 개선했습니다.

isExample committed:
[Refactor] HTTP Status Code 400에서 404로 변경

서버가 client 요청 자원을 가지고 있지 않는 경우인 404가 더 적합하여 변경했습니다.

isExample committed:
[Refatcor] mapper 메서드명 select~에서 find~로 변경 & 키워드 중복 제거

특정 필드값을 기반으로 특정 레코드를 찾아온다는 점에서 find~가 더 적절한 것 같아 변경했습니다.
또한 기존에는 select'Member'ById와 같이 객체명이 불필요하게 중복되어 제거했습니다.

isExample committed:
[Refactor] 변경된 mapper 메서드명 반영

service 계층에서 사용되는 mapper 메서드명에 변경 사항을 반영했습니다.

isExample committed:
[Feature] devtools 의존성 추가

Automatic Restart, Live Reload를 통해 변경 요소를 저장하면 자동으로 서버를 재시작하도록 설정했습니다.

isExample committed:
[Feature] Gradle Wrapper 구성

Gradle Wrapper를 생성하고 버전 Url을 지정하여 구성했습니다.

isExample committed:
[Refactor] Controller에서의 HttpSession 의존성 제거 & 세션 속성 Id로 변경

SessionService를 만들어 Member 관련 controller에서의 HttpSession 의존성을 제거했습니다.
또한, session에 회원 이메일이 아닌 Id를 저장하도록 개선했습니다.

isExample committed:
[Refactor] HttpSession에 저장하는 회원 정보 객체화

HttpSession에 저장하는 회원 정보를 확장성과 명확성을 위해 객체로 생성했습니다.

isExample committed:
[Feature] 로그인 회원 인증을 위한 Interceptor와 Argument Resolver 구현

로그인 회원 인증 로직의 중복을 피하기 위해 Interceptor와 Argument Resolver를 구현했습니다.

isExample committed:
[Refactor] 로그인 인증 Interceptor와 Resolver에 주석 추가

간단한 주석을 추가헸습니다.

isExample committed:
[Feature] Interceptor와 Argument Resolver를 MVC 설정에 등록

로그인 인증을 위한 커스텀 Interceptor와 Argument Resolver를 등록했습니다.

isExample committed:
[Refactor] 사용하지 않는 클래스들 제거

Commit에 실수로 포함한 미완성 코드들을 제거합니다.

isExample committed:
[Refactor] RequestBody DTO immutable하도록 개선

RequestBody DTO를 명시적으로 Immutable하게 개선했습니다.

isExample committed:
[Refactor] 불필요한 import문 제거

사용하지 않는 패키지 import문을 제거했습니다.

isExample commented:
@flab-true 멘토님이 작성해 주신 리뷰들은 모두 확인하고 답변 남겼습니다. 우선 다음 구현을 진행하기 위해 merge 진행하겠습니다. 추가된 커밋이나 그 외의 사항에 대한 피드백은 언제든 남겨주시면 확인하고 별도의 PR로 다뤄보도록 하겠습니다. 감사합니다!
